hippy/module/pypy_bridge/py_strategies.py

```diff
@@ -97,6 +97,33 @@ def pop(self, w_list, index):
 
         return w_php_val.to_py(interp)
 
+    def getslice(self, w_list, start, stop, step, length):
+        """Adapted from the PyPy implementation"""
+
+        interp = self.space.get_php_interp()
+        w_php_arry_ref = self.unerase(w_list.lstorage)
+        w_php_arry = w_php_arry_ref.deref_temp()
+
+        if step == 1 and 0 <= start <= stop:
+            assert start >= 0
+            assert stop >= 0
+            sublist = w_php_arry._get_raw_storage()[start:stop]
+            sublist_py = [x.to_py(interp) for x in sublist]
+
+        else:
+            sublist_py = [self._none_value] * length
+            for i in range(length):
+                try:
+                    sublist_py[i] = \
+                        w_php_arry._get_raw_storage()[start].to_py(interp)
+                    start += step
+                except IndexError:
+                    raise
+
+        # desired strategy is unknown, so create via objspace
+        return self.space.newlist(sublist_py)
+
+
```

Comment:
Yes, the question is whether the list is an adapted Python list, or a native one. I have no strong opinion regarding which is correct, other than my reasoning above. And I could be convinced otherwise.

Regarding efficiency, I devised a rough benchmark to see how much slowdown the current implementation incurs:

First a baseline array slice program in PHP (slice_native.php):

```
function do_slice($a) {                                                         
        $l = count($a);                                                         
        $tot = 0;                                                               
        for ($i = 0; $i < $l; $i++) {                                           
                // equiv to a[i:]                                               
                $slice = array_slice($a, $i);                                   
                $tot += array_sum($slice);                                      
        }                                                                       
        return $tot;                                                            
}                                                                               
                                                                                
$ary = range(0, 999); // inclusive upper                                        
for ($j = 0; $j < 100; $j++) {                                                  
        do_slice($ary);                                                         
}
```

And one where the slice is in Python (slice_comp.php):

```
$pysrc = <<<EOD                                                                 
def sl(a, s):                                                                   
        return a.as_list()[s:]                                                  
EOD;                                                                            
compile_py_func_global($pysrc);                                                 
                                                                                
function do_slice($a) {                                                         
        $l = count($a);                                                         
        $tot = 0;                                                               
        for ($i = 0; $i < $l; $i++) {                                           
                $slice = sl($a, $i);                                            
                $tot += array_sum($slice);                                      
        }                                                                       
        return $tot;                                                            
}                                                                               
                                                                                
$ary = range(0, 999); // inclusive upper                                        
for ($j = 0; $j < 100; $j++) {                                                  
        do_slice($ary);                                                         
}
```

I've checked that both benchmarks yield the same total.

Some rough timings:

```
vext01@bencher2:~/research/pyhyp/hippyvm.hippyvm$ /opt/multitime/bin/multitime -n 5 ./hippy-c /tmp/slice_native.php 
===> multitime results
1: ./hippy-c /tmp/slice_native.php
            Mean        Std.Dev.    Min         Median      Max
real        5.109       0.027       5.085       5.089       5.149       
user        5.092       0.024       5.068       5.076       5.128       
sys         0.008       0.004       0.004       0.008       0.012       
vext01@bencher2:~/research/pyhyp/hippyvm.hippyvm$ /opt/multitime/bin/multitime -n 5 ./hippy-c /tmp/slice_comp.php 
===> multitime results
1: ./hippy-c /tmp/slice_comp.php
            Mean        Std.Dev.    Min         Median      Max
real        1.259       0.027       1.240       1.247       1.313       
user        1.248       0.028       1.224       1.236       1.300       
sys         0.008       0.007       0.000       0.008       0.020       
```

The composed program is faster, even with `array_sum` working over a non-native array. I guess because PyPy lists are optimised better than Hippy arrays. I didn't expect this to be honest, but it does diffuse the idea that it would be faster to keep the array a PHP array instead of an adapted Python array (?). Is my logic correct?